### PR TITLE
Authx - Add in new JWT Entity with Create and Validate methods to permi…

### PR DESCRIPTION
**Pull request by seamuslee001**
…t creation and validation of JWTs via the API

Overview
----------------------------------------
This PR adds in JWT.create and JWT.validate actions to allow 3rd party applications to potentially generate JWTs over APIs when you're in a service principle concept

Before
----------------------------------------
No Methods to create JWT or validate it over the API

After
----------------------------------------
New Methods to do the above

ping @totten @johntwyman

### CODE REVIEW

File: ext/authx/tests/phpunit/api/v4/JwtApiTest.php [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    $this->expectExceptionMessage('Expired token');
+    $this->_apiversion = 4;
+    $contactRecord = $this->createTestRecord('Contact', ['contact_type' => 'Individual']);
+    $this->createLoggedInUser();
+    $this->setPermissions([
+      'access CiviCRM',
+      'generate JWT',
+    ]);
+    $jwt = JWT::create()->setContactId($contactRecord['id'])->setTtl(5)->execute();
+    sleep(10);
```
Comment (colemanw) [inline ext/authx/tests/phpunit/api/v4/JwtApiTest.php:76]: The test would be 10 seconds faster without this :/

Comment (seamuslee001) [reply ext/authx/tests/phpunit/api/v4/JwtApiTest.php:76]: Yes well it needs to slow down at least 5s because i'm waiting for the JWT to expire

**Comment by totten** [comment]
### r-run

For testing purposes, I've used this setup:

```bash
drush -y role-create auth-role
drush -y role-create validate-role
drush -y rap auth-role 'access CiviCRM,generate JWT'
drush -y rap validate-role 'access CiviCRM'
drush -y user-create --password="demo" --mail="auth-user@example.com" "auth-user"
drush -y user-create --password="demo" --mail="validate-user@example.com" "validate-user"
drush -y user-add-role auth-role auth-user
drush -y user-add-role validate-role validate-user
```azz

And tried a few commands:

```bash
echo 'http://dmaster.bknix:8001/civicrm/dashboard?_authxSes=1&_authx=Bearer+'$(cv api4 -IU auth-user JWT.create contactId=203 checkPermissions=1)
## Makes a valid login link

echo 'http://dmaster.bknix:8001/civicrm/dashboard?_authxSes=1&_authx=Bearer+'$(cv api4 -IU validate-user JWT.create contactId=203 checkPermissions=1)
## Fails with permission error, as expected

cv api4 -U validate-user JWT.validate checkPermissions=1 token=$(cv api4 -IU auth-user JWT.create contactId=203 checkPermissions=1)
## Fails: "Class 'Civi\Authx\AuthenticatorTarget' not found"
## If you comment-out the call to `AuthenticatorTarget::create`, it nominally runs, but doesn't work due to
## [PHP Notice] Undefined offset: 1 at /home/totten/bknix/build/dmaster/web/sites/all/modules/civicrm/ext/authx/Civi/Authx/CheckCredentialEvent.php:54

cv api4 -U validate-user JWT.validate checkPermissions=1 token=Bearer\ $(cv api4 -IU auth-user JWT.create contactId=203 checkPermissions=1)
## Note: This manually adds the "Bearer" prefix.
## Fails: "Class 'Civi\Authx\AuthenticatorTarget' not found"
## If you comment-out the call to `AuthenticatorTarget::create`, it works.
```

### Design

I feel a bit of ambiguity in the design of the signatures. Maybe think of it this way -- there are two levels of functionality:

* __(Lower level) Signed Message__ - Create a signed-message. Later, validate the signed message. The payload (*scopes and other claims*) are circumstantial/flexible/open-ended.

* __(Higher level) Login Credential__ - Create a login-credential. Later, validate the login-credential. The credential could be a JWT (or something else). For  a JWT credential, the payload (scopes+credentials) are dictated by the login system.

Depending on which level of functionality one aims for, the `create` and `validate` methods would have different contracts and implementations. Roughly:

* __(Lower level) Signed Message__ - The `create` should allow an open-ended payload. The `validate` would return the validated data (regardless of the payload).  This shouldn't need `AuthenticatorTarget` or `CheckCredentialsEvent`. It just needs `crypto.jwt` service. The API-consumer is responsible for defining and using the payload. The `Bearer` prefix would not appear anywhere.

* __(Higher level) Login Credential__ - The payload (*scopes and other claims*) and the use of JWT are internal details. `create` just takes  the principal and TTL. `validate` just returns the validated principal. This might use things like `AuthenticatorTarget` and `CheckCredentialsEvent`.

It looks to me like the current implementation is leaning more toward the high-level (*login credential*)? If that is the goal, then...

* The entity name `AuthxCredential` is probably better than `JWT`.
* For `Create`, AFAICS, there's really nothing you can do with `$scope`. I'd like to see more discussion around concrete use-cases (or maybe    hold-off on `$scope` until we have some).
* For `Validate`, we should think a bit more about it interacts with `flow`, `cred`, and `Authenticator`. (*But I'll hold off on that until you've confirmed that the intent is more like _login credential_.*)

**Comment by seamuslee001** [comment]
@totten yeh I think I want more of a login / authorization credential, perhaps the lower level one could be its own method, maybe we make this CreateLoginCredential / ValidateLoginCredential or something like that?

**Comment by JoeMurray** [comment]
I think it is likely helpful to have both higher and lower level access, but want to push folks to consider if this opens up any issues with security when an programmed interaction based on higher level login interface might be susceptible to an attack if the lower level API is not private, for example, changing the scope. Can we programmatically preclude any token for login ever being acceptable for changing scopes, etc. The source could be forked and changed, somehow applied to instance, etc. in highly unlikely but possible attack... (if they can do that then they could also more easily..., so, not pushing this as reasonable scenario, just pushing you developers closer to the implementation to try to think of problems.)

**Pull request merged**